app/controllers/admin/institutions_controller.rb

```diff
@@ -1,64 +1,69 @@
+  before_action :set_action_page, only: [:edit_targets, :update_targets]
+  before_action :set_institution, only: [:destroy, :edit, :update]
 
-  require "csv"
-
-  # GET /admin/action_pages/:action_page_id/institutions
   def index
-    @institutions = @actionPage.institutions.order(:name).page(params[:page])
+    @institutions = Institution.all
+    @institutions = @institutions.search(params[:q]) if params[:q].present?
+    @institutions = @institutions.paginate(page: params[:page], per_page: 20)
   end
 
-  # GET /admin/action_pages/:action_page_id/institutions/new
   def new
-    @institution = @actionPage.institutions.new
+    @institution = Institution.new
   end
 
-  # POST /admin/action_pages/:action_page_id/institutions
   def create
-    @institution = Institution.find_or_initialize_by(name: institution_params[:name])
-    @actionPage.institutions << @institution
-
-    respond_to do |format|
-      if @institution.save
-        format.html { redirect_to [:admin, @actionPage, Institution], notice: "Institution was successfully created." }
-      else
-        format.html { render "new" }
-      end
+    @institution = Institution.find_or_initialize_by(institution_params)
+
+    if @institution.save
+      redirect_to "index", notice: "#{@institution.name} was successfully created."
+    else
+      render "index"
     end
   end
 
-  # POST /admin/action_pages/:action_page_id/institutions/import
-  def import
-    names = []
-    CSV.foreach(params[:file].path, headers: true) do |row|
-      params = row.to_hash
-      unless params["name"]
-        flash[:notice] = "Import failed. Please check CSV formatting"
-        break
-      end
-      names << params["name"]
+  def edit; end
+
+  def update
+    if @institution.update(institution_params)
+      redirect_to "index", notice: "#{@institution.name} was successfully updated."
+    else
+      render "edit"
     end
+  end
 
-    Institution.delay.import(names, @actionPage) if names.present?
+  def upload; end
 
-    redirect_to [:admin, @actionPage, Institution]
+  def import
+    names = Institution.process_csv(params[:file])
+    if names.empty?
+      redirect_to "upload", notice: "Import failed. Please check CSV formatting"
+    else
+      Institution.delay.import(import_params[:category], names)
+      redirect_to "index", notice: "Successfully imported #{names.length} targets"
+    end
   end
 
-  # DELETE /admin/action_pages/:action_page_id/institutions/1/
   def destroy
-    @actionPage.institutions.delete(@institution)
-    respond_to do |format|
-      format.html { redirect_to [:admin, @actionPage, Institution] }
-    end
+    @institution.destroy
+    redirect_to "index", notice: "Target deleted"
   end
 
-  # DELETE /admin/action_pages/:action_page_id/institutions/
-  def destroy_all
-    @actionPage.institutions.delete_all
-    respond_to do |format|
-      format.html { redirect_to [:admin, @actionPage, Institution] }
-    end
+  def edit_targets
```

Comment:
this and `update_targets` might actually fit better in a new controller (`Admin::ActionInstitutionsController`) because it deals with creating and removing those join models.